File: src/main/java/com/example/kotlinDemo/java/classes/Person.java
package com.example.kotlinDemo.java.classes;

public class Person {
    private final String name;
    private String nickName;
    private Boolean isMarried;
    private final String upperCaseName;

    public Person(String name, String nickName, Boolean isMarried) {
        this.name = name;
        this.nickName = nickName;
        this.isMarried = isMarried;
        this.upperCaseName = name.toUpperCase();
    }

    public String getName() {
        return name;
    }

    public String getNickName() {
        return nickName;
    }

    public void setNickName(String nickName) {
        this.nickName = nickName;
    }

    public Boolean getMarried() {
        return isMarried;
    }

    public void setMarried(Boolean married) {
        isMarried = married;
    }

    public String getUpperCaseName() {
        return upperCaseName;
    }

    @Override
    public String toString() {
        return "Person{" +
                "name='" + name + '\'' +
                ", nickName='" + nickName + '\'' +
                ", isMarried=" + isMarried +
                ", upperCaseName='" + upperCaseName + '\'' +
                '}';
    }
}

File: src/main/kotlin/com/example/kotlinDemo/kotlin/classes/Person.kt
package com.example.kotlinDemo.kotlin.classes

class Person(
    val name: String,
    var nickName: String?,
    var isMarried: Boolean
) {
    val upperCaseName = name.uppercase()

    fun getDescription(): String {
        return "$name is married: $isMarried"
    }

    override fun toString(): String {
        return "Person(name='$name', nickName=$nickName, isMarried=$isMarried, upperCaseName='$upperCaseName')"
    }
}

//open class PersonWithDefault(
//    val name: String,
//    var nickName: String? = null,
//    var isMarried: Boolean = false
//) {
//
//    val upperCaseName = name.uppercase()
//
//    fun getNicknameIfExists(): String {
//        return nickName ?: "nick name is not set"
//    }
//
//    override fun toString(): String {
//        return "Person(name='$name', nickName=$nickName, isMarried=$isMarried, upperCaseName='$upperCaseName')"
//    }
//}

//class PersonWithDetails(name: String, nickName: String? = null, isMarried: Boolean = false, val placeOfBirth: String) : PersonWithDefault(name, nickName, isMarried) {
//
//    override fun toString(): String {
//        return "PersonWithDetails(name='$name', nickName=$nickName, isMarried=$isMarried, upperCaseName='$upperCaseName, placeOfBirth='$placeOfBirth')"
//    }
//}